storage/badger/consumer_progress.go

```diff
@@ -22,26 +67,15 @@ func NewConsumerProgress(db *badger.DB, consumer string) *ConsumerProgress {
 
 func (cp *ConsumerProgress) ProcessedIndex() (uint64, error) {
 	var processed uint64
-	err := cp.db.View(operation.RetrieveProcessedIndex(cp.consumer, &processed))
+	err := operation.RetrieveProcessedIndex(cp.consumer, &processed)(operation.ToReader(cp.db))
 	if err != nil {
 		return 0, fmt.Errorf("failed to retrieve processed index: %w", err)
 	}
 	return processed, nil
 }
 
-// InitProcessedIndex insert the default processed index to the storage layer, can only be done once.
-// initialize for the second time will return storage.ErrAlreadyExists
```

Comment:
With batch updates, it's not easy to prevent dirty reads in this call when check the case that would return ErrAlreadyExists. We would need to introduce a lock. 

Instead of using lock, I decided to move this check to a Factory struct, since it's easy for the owner of the Factory to ensure initializing in a single thread, and then the returned consumer progress could be concurrent-safe.